Merge 15.0 back into 15.0.project_PR_449

BT-jcolmeiro opened the pull request.

bobslee committed:
[UPD] formio_storage_filestore: installable

Tested and already works.

bt-admin committed:
Merge pull request #80 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[FIX] formio: Major improvements for translations

Major improvements for translations:
- Specific Version Translations instead of linking (Many2many) to the available Base Translations.
- Translations (overrides) of formio.js source properties in the form builder.

Add `noupdate=1` for the xmlid `formio.version_dummy` data (record).
This prevents recreation when the dummy version has been archived (is inactive).

bobslee committed:
[IMP] formio: Minor version translation improvements.

bobslee committed:
[IMP] formio: version translation view improvement.

bobslee committed:
[15.0][MIG] formio: backport migration 16.0.2.0

This relates to previous commit c6b918128e86fa99632b0ec739b70892c9eb5b6a

bobslee committed:
[IMP] formio: Version translation improvements.

bobslee committed:
[IMP] formio: Version translation improvements.

Improvements for Version Translations (model: `formio.version.translation`):
- Compute and store origin (base translation, user added).
- Compute and show whether a copied (origin) base translation has been
  updated.
- Add sequence field. So translations can be ordered to ease
  admnistration and the translation override implementation.

Fix `name_get` method (models: `formio.translation`, `formio.version.translation`)

bobslee committed:
[FIX] formio: Remove unused compute arg from field.

bt-admin committed:
Merge pull request #81 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[FIX, SECURITY] formio_storage_filestore

Fix security vulnerability for GET (file) request.
The form access-check was missing here, which determines whether
the (file) request is allowed or forbidden.

Mairenhub committed:
[IMP] formio: 'Forms' Button in the Form Builder view.

bt-admin committed:
Merge pull request #84 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
website_formio: Rename demo title 'formio_builder_demo_website'

bobslee committed:
[FIX] formio: formio.js version 'Download and install' ValueError

ValueError: max() arg is an empty sequence.
Initialize the translations sequence properly.

bt-admin committed:
Merge pull request #87 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio_storage_filestore: Improve public user check

Improve public user check, by checking either:
- no user in request.env, or ...
- the request user has the public user group (the added check).
This change isn't solving a vulnerability, but a public GET request
was too restrictive and could be denied too quickly.

bobslee committed:
[IMP] formio: Migration 15.0.13.0 formio.version translation_ids

bobslee committed:
[DOC] formio_storage_filestore: CHANGELOG index.html

bobslee committed:
[DOC] formio_storage_filestore: Typo index.html

bt-admin committed:
Merge pull request #92 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio: Change form 'heightCalculationMethod' to 'taggedElement'.

This solves issues for components with dynamic height, eg the collapsible Edit Grid (type `editgrid`).

bobslee committed:
[IMP] formio: formio.version.translation base_translation_updated (reaodnly).

bobslee committed:
[FIX] formio: Add heightCalculationMethod targets in portal and public templates.

bobslee committed:
Add field deprecation comment.

bobslee committed:
[UPD] Update iframe-resizer v4.3.6

bt-admin committed:
Merge pull request #96 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[DOC] formio: index.html (info).

bobslee committed:
[DOC] formio: index.html (info).

bobslee committed:
[DOC] formio: index.html (minor).

bobslee committed:
[DOC] formio: index.html (info).

bobslee committed:
[DOC] formio: index.html (info).

bobslee committed:
[DOC] formio: index.html (typo).

bobslee committed:
[DOC] formio_storage_filestore: index.html (header).

bobslee committed:
[DOC] formio_storage_filestore: name in manifest

bobslee committed:
[DOC] formio: index.html (minor).

bt-admin committed:
Merge pull request #100 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[FIX] formio: Fix language determination (cascade) for public Form.

- Fix language determination (cascade) for public Form load by UUID.
- Add cascade delete on `formio_version_id` in model
  `formio.version.translation`.
- Improve code readability in formio.form method _public_access .

bt-admin committed:
Merge pull request #104 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[DOC] formio_storage_filestore: Rename module title (more obvious).

bobslee committed:
[IMP] formio_crm, formio_sale: Add cascade delete on Many2one in form models

- formio_crm: cascade delete `crm_lead_id` field in `formio.form` model.
- formio_sale: cascade delete `sale_order_id` field in `formio.form` model.

bobslee committed:
[DOC] formio_storage_filestore: Rename module title (more obvious).

bt-admin committed:
Merge pull request #107 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio: Form Builder conditional api_alert (warning)

In the Form Builder, implement conditional warning (also extendable)
in case there are integrated APIs.
This warns the Form Builder user about components being removed or
updated, which could lead the API regression/breakage.

bt-admin committed:
Merge pull request #109 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] website_formio: Solve Form height resizing issues

Apply the iFrame Resizer `bodyMargin` option.

bobslee committed:
[IMP] formio: Improve public/website languages determination.

bobslee committed:
[IMP] formio: Major - replace _onchange_builder_id method. Improve state usage.

1. Refactored the Form its Form Builder (field) domain/filter.

This solves the deprecation warning, regarding a domain that may not be returned by an onchange method.

Replace the `_onchange_builder_id` method (implementation), by a new computed field `builder_id_domain` and 2 new methods:
- `_compute_builder_domain`
- `_get_builder_domain`.

This also affects the implementation replacement in additional modules which should be updated as well !
Eg: `formio_crm`, `formio_partner`, `formio_purchase`, `formio_sale`.

**WARNING / UPDATE REQUIREMENT**

Update all modules which implemented the `_onchange_builder_id` method.
For the (Nova Code) Forms modules those are:
- formio_crm
- formio_partner
- formio_purchase
- formio_sale

2. Add new feature which allows Form Builders in "Draft" or "Obsolete" state to be choosen in a new Form in the backend.

The 2 setting fields are available in the Form Builder:
- Use Draft in Backend
- Use Obsolete in Backend

bobslee committed:
[IMP] formio: Builder form-view, add invisible for backend_use_* fields.

bobslee committed:
[IMP] formio: Builder form-view, layout backend_use_* fields.

bobslee committed:
[IMP] formio: Builder form-view, layout backend_use_* fields.

bobslee committed:
[IMP] formio: Improve app/module installation

Download and install a default formio.js library version after install.
Also log a warning in case an error (eg ConnectionError) occurs.

bt-admin committed:
Merge pull request #113 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[FIX] formio: Add migrations for version 16.0.3.1, fix for:

UniqueViolation: duplicate key value violates unique constraint "ir_config_parameter_key_uniq"

bobslee committed:
[IMP] formio: Some info in Form Builder form-view about formio.js version

bobslee committed:
[FIX] formio: Port migrations to version 15.0.14.1, fix for:

UniqueViolation: duplicate key value violates unique constraint "ir_config_parameter_key_uniq"

bobslee committed:
[FIX] formio: post-migration 16.0.3.1 query 'not installed'

Only download and install the default formio.js version, if it's not
installed yet.

bobslee committed:
[IMP] formio: archive functionality Formio.js versions

Added the archive functionality in the form of a toggle button for formio.js versions.
Also added predefined filters to filter for archived versions.

Close #126

bobslee committed:
[IMP] formio: Call form's beforeSubmit in wizard draft autosave mode

bobslee committed:
[IMP] formio: language buttons styling improvements.

bobslee committed:
[IMP] formio: Fix migration. Several improvements.

Fix
---

Fix migration for version 15.0.14.2 - upgrade error:
UniqueViolation: duplicate key value violates unique constraint "ir_config_parameter_key_uniq"

Improvements
------------

Style (highlight) active language button.

The "Forms" Button in the Form Builder view.

Archive functionality for formio.js versions (also unarchive and search filter).

Wizard draft autosave mode.
This allows complex components to serialize their data on the form before submitting it.

bobslee committed:
[IMP] formio: Improve Forms button in Form Builder.

bobslee committed:
[IMP] formio: title field (length) in Form Builder and Form.

bt-admin committed:
Merge pull request #116 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio: Update colors language buttons to light grey.

bobslee committed:
[IMP] formio: Styling of Form Builder button Fullscreeen.

bobslee committed:
[IMP] formio: Styling (2) of Form Builder button Fullscreeen.

bt-admin committed:
Merge pull request #118 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] website_formio: is_published field in view.

sjamaan committed:
Make formio.js CDN for hot-loaded dependencies configurable

Make use of the formio.js CDN class to override the base URL for
all 3rd party assets that are loaded on-the-fly by formio.js lib.
This includes for example ACE, CKEditor, Flatpickr, Quill etc.

This allows for improved privacy by using a CDN that honors the GDPR.
Since I was unable to find any information about GDPR compliance of
the default formio.js CDN, I've set the Cloudflare CDN as the default
value.  Even though it's not perfect, they have at least a page about
the GDPR: https://www.cloudflare.com/trust-hub/gdpr/

Commercial sites in the EU could choose to override the default value
to a paid GDPR-aware CDN like KeyCDN.com or GlobalConnect.no and host
the required files there, or simply point it to the base URL of their
Odoo install and ship the files themselves.  This would require
pinning the formio.js version, since different versions of the library
need different dependencies.

sjamaan committed:
Patch the CDN class's buildUrl for ckeditor

Apparently, ckeditor's versioning is a bit iffy - formio.js tries to
load ckeditor 19.0.0, but that's actually ckeditor 5 v19.0.0.  Also,
19.0.0 is not available at Cloudflare's CDN so we substitute v19.0.1
which *is* available.

Unfortunately, the CDN class is not exported by formio.js so we can't
subclass it properly.  Instead, just monkey patch the method.

sjamaan committed:
Also override flatpickr to avoid custom formio.js version

The library 'flatpickr-formio' is not typically available on public
CDNs.  Strangely, this seems to be a hacked version of the regular
flatpickr, because when you pre-load flatpickr statically, the CDN
will not attempt to load flatpickr-formio either.

bobslee committed:
[DOC] formio: Changelog 15.0.15.0

bobslee committed:
[IMP] formio: Implement fallback in patchCDN - formio.js <= 4.14.12

The patch implements a fallback for formio.js version <= 4.14.12, where
CDN.buildUrl is not implemented, to patch CDN.updateUrls.

bobslee committed:
[DOC] formio: Changelog 15.0.15.0

bt-admin committed:
Merge pull request #121 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[DOC] formio: Changelog 15.0.15.0

bt-admin committed:
Merge pull request #123 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio: cdn_base_url for portal and public forms.

Add cdn_base_url to formio.builder methods:
- _get_portal_form_js_params
- _get_public_form_js_params

bobslee committed:
[FIX] formio: Backwards formio.js patchCDN compatibility fix

Apply `patchCDN` if there's an instantiated `Formio.cdn` object.

bobslee committed:
[FIX] formio: Remove form builder CDN fallback.

bobslee committed:
[IMP] formio: Render a new portal form by the generic QWeb form template.

This ensures to apply any extra form builder config/settings.

bobslee committed:
[DOC] formio: Improve changelog 15.0.14.0

bt-admin committed:
Merge pull request #126 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[FIX] formio: Wrong (copied) version in __manifest__.py

bobslee committed:
[IMP] formio: Patch the formio.js requireLibrary to alter some URLs.

Ensure external JavaScript libraries get properly loaded (by the formio.js `CDN` class).\
This introduces the `patchRequireLibrary` (JS function) which alters some library URLs a common CDN expects.\
Currently the Flatpickr localizations URLs are altered.

bt-admin committed:
Merge pull request #128 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[FIX] formio: Fix and improve 'datetime' component localization and translations

- Fix and improve `patchRequireLibrary` (JS function) to gradually determine the arguments.
- Fix and improve Form Builder `setLanguage` to localize datetime components.

bt-admin committed:
Merge pull request #130 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio: Implement `api=getData` model access-rights bypass (by ir.rule).

Implement a `api=getData` model access-rights bypass, by obtaining a direct record rule (`ir.rule`).
Provide the  record-rule (`ir.rule`) ID in the query-param `&module_rule_id=<model('ir.rule'):model_rule_id>`, shown in the example below.
The domain of the record-rule applies and the model `search_read` shall be executed with superuser mode (`sudo()`).

Example URL (query params):
`http://localhost:8069/formio/portal/form/new/FORMIO_BUILDER_NAME?api=getData&model=stock.production.lot&label=name&model_rule_id=99`

bt-admin committed:
Merge pull request #132 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio: Improve `api=getData` model access:

- Also obtain `ir.rule` external ID in the URL's query-string.
- Groups check to harden/scope access.
- Only add domain if can be determined.

bt-admin committed:
Merge pull request #133 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio: ir.actions.server add eval context, `os.getenv` and `logger`

bt-admin committed:
Merge pull request #134 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio: Implement validation hook. Rename form payload (data to submission).

Changes:
- Implement the formio.js `customValidation` hook, which XHR posts to
  the API /validation endpoint.
- Rename form payload from `data` to `submission`, which is more
  self-explanatory.
- Some cleanup and comments.

bobslee committed:
[DOC] formio: release version 15.0.15.7

bobslee committed:
[DOC] formio: release version 15.0.16.0 (changelog)

bobslee committed:
[DOC] formio: minor typo changelog.

bobslee committed:
Merge pull request #247 from bobslee/15.0-validation-api

15.0: Validation API hook

bobslee committed:
[FIX] formio: Other renames of form payload (data to submission).

bobslee committed:
[IMP] formio: Renaming submission / form_data.

bt-admin committed:
Merge pull request #140 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

BT-dmoreno reviewed (approved):
I've reviewed the changes and, there's a problem with forms on load, that prevents buttons to work, or select fields to be used neither. That's because we need to update a method in bt_formio module. After that change, probably everything should be fine, since I don't see any other potential problem at the moment.

Apart from comments, and cosmetic changes, the main points of this update seems to be:

- The use of alternative way of checking for rule accesses in the options of select components (we don't have nothing directly hooked here, so OK).
- New model for Formio version translations (our current translations still should work anyway).
- iFrameResizer seems to have been updated.

bobslee committed:
[FIX] formio: Standalone `formio.form` create, shouldn't set `res_id`.

Fix in with context check in function `_prepare_create_vals`.

bt-admin committed:
Merge pull request #142 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[FIX] formio_component_recaptcha: Include the assets directly in template XMLs.

Include the assets (files) directly in the template XMLs.
Remove `formio.extra.asset` data.

bt-admin committed:
Merge pull request #144 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio: Form Builder form-view, add group "main" in page "formio_components_api".

bobslee committed:
[IMP] formio_components_synchronizer migration and improvements:

- Add fields to the Component record, to show the (full, input) paths with keys and labels in the related Form Builder.
- Distinguish input and layout components. Therefor add `formio.component` boolean field `input`.
- Other improvements.

bt-admin committed:
Merge pull request #145 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)

bobslee committed:
[IMP] formio: Improvement for translations:

- Add translate function in `formio.builder` model
- Set language (IETF code) in payload of `customValidation` hook.

bt-admin committed:
Merge pull request #147 from novacode-nl/15.0

Syncing from upstream novacode-nl/odoo-formio (15.0)
